pallets/settlement/src/lib.rs

```diff
@@ -1320,76 +1368,105 @@ impl<T: Trait> Module<T> {
+                                        &to_mercat_balance, // If the only reason we need this is updating balance, we should remove it.
+                                        &[],
+                                        &mut rng,
                                     )
-                                {
-                                    if let Ok((to_mercat, to_mercat_balance)) =
-                                        Self::get_account_and_balance(
-                                            &leg.to.did,
-                                            &leg.to_account_id,
-                                        )
-                                    {
-                                        // Verify the proofs.
-                                        let mut rng = rng::Rng::default();
-                                        let result = TransactionValidator {}.verify_transaction(
-                                            &decoded_justified_tx,
-                                            &from_mercat,
-                                            &from_mercat_balance,
-                                            &to_mercat,
-                                            &to_mercat_balance,
-                                            &[],
-                                            &mut rng,
-                                        );
-                                        if let Ok((new_sender_balance, new_receiver_balance)) =
-                                            result
-                                        {
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
+                                    .and_then(|_| {
+                                        // Note that storage failures could be a problem, if only some of the storage is updated.
+                                        let _ =
+                                            <ConfidentialAsset<T>>::mercat_account_withdraw_amount(
                                                 &leg.from.did,
                                                 &leg.from_account_id,
-                                                new_sender_balance,
-                                            );
-                                            <ConfidentialAsset<T>>::update_mercat_account_balance(
+                                                decoded_justified_tx
+                                                    .finalized_data
+                                                    .init_data
+                                                    .memo
+                                                    .enc_amount_using_sndr,
+                                            )
+                                            .map_err(|_| (leg_id, FailureReason::GetMercatAccount));
+
+                                        let _result =
+                                            <ConfidentialAsset<T>>::mercat_account_deposit_amount(
```

Comment:
Something is odd here. The `mercat_account_deposit_amount` is changing the pending state. So I was expecting to see it being called in the `authorizeConfidentialInstruction` function. Similarly for `mercat_account_withdraw_amount`.

This piece of code is called when the transaction is justified, which means that in this code we should do cleanup of the pending state as you are doing by calling `remove_tx_pending_state` and `add_failed_outgoing_balance`

Reply:
Yes. On transaction validation success, we deposit/withdraw assets from the receiver/sender on-chain account. In case of the sender, we subtract the transaction's balance from the `pending_outgoing` accumulator, and their on-chain balance. In case of the receiver, we add the transaction's balance to the `incoming_balance` accumulator, and their on-chain balance.

`authorize_confidential_instruction()` updates sender's pending state when they first add the transaction to the instruction. Does that answer your question?

Reply:
Yep